ipyparallel/cluster/shellcmd.py

```diff
@@ -0,0 +1,859 @@
+            )
+            print(f'__remote_pid={p.pid}__')
+            sys.stdout.flush()
+
+    def cmd_running(self, pid):
+        self._log(f"Check if pid {pid} is running")
+
+        if self.platform == Platform.Windows:
+            # taken from https://stackoverflow.com/questions/568271/how-to-check-if-there-exists-a-process-with-a-given-pid-in-python
+            import ctypes
+
+            PROCESS_QUERY_INFROMATION = (
+                0x1000  # if actually PROCESS_QUERY_LIMITED_INFORMATION
+            )
+            STILL_ACTIVE = 259
+            processHandle = ctypes.windll.kernel32.OpenProcess(
+                PROCESS_QUERY_INFROMATION, 0, pid
+            )
+            if processHandle == 0:
+                print('__running=0__')
+            else:
+                i = ctypes.c_int(0)
+                pi = ctypes.pointer(i)
+                if ctypes.windll.kernel32.GetExitCodeProcess(processHandle, pi) == 0:
+                    print('__running=0__')
+                if i.value == STILL_ACTIVE:
+                    print('__running=1__')
+                else:
+                    print('__running=0__')
+                ctypes.windll.kernel32.CloseHandle(processHandle)
+        else:
+            try:
+                # use os.kill with signal 0 to check if process is still running
+                os.kill(pid, 0)
+                print('__running=1__')
+            except OSError:
+                print('__running=0__')
+
+    def cmd_kill(self, pid, sig=None):
+        self._log(f"Kill pid {pid} (signal={sig})")
+
+        if self.platform == Platform.Windows:
+            # os.kill doesn't work reliable under windows. also see
+            # https://stackoverflow.com/questions/28551180/how-to-kill-subprocess-python-in-windows
+
+            # solution using taskill
+            # import subprocess
+            # subprocess.call(['taskkill', '/F', '/T', '/PID',  str(pid)])  # /T kills all child processes as well
+
+            # use windows api to kill process (doesn't kill children processes)
+            # To kill all children process things are more complicated. see e.g.
+            # http://mackeblog.blogspot.com/2012/05/killing-subprocesses-on-windows-in.html
+            import ctypes
+
+            PROCESS_TERMINATE = 0x0001
+            kernel32 = ctypes.windll.kernel32
+            processHandle = kernel32.OpenProcess(PROCESS_TERMINATE, 0, pid)
+            if processHandle:
+                kernel32.TerminateProcess(
+                    processHandle, 3
+                )  # 3 is just an arbitrary exit code
+                kernel32.CloseHandle(processHandle)
+        else:
+            os.kill(pid, sig)
+
+    def cmd_mkdir(self, path):
+        self._log(f"Make directory '{path}'")
+
+        os.makedirs(path, exist_ok=True)  # we allow that the directory already exists
+
+    def cmd_rmdir(self, path):
+        self._log(f"Remove directory '{path}'")
+
+        import shutil
```

Comment:
Thanks, that makes sense. I think it might make the whole thing more manageable if we make the unit we send a whole standalone script, rather than fragments of this file.

Reply:
You mean sending the whole _shellcmd.py_ script or splitting it into two files? I guess keeping everything in one place is beneficial and sending a bit to much code doesn't hurt, right?

Reply:
Thanks for breaking down everything so nicely!
The zero-dependency requirement on the sender part is quite a strong argument against a single file send (although currently, there is no dependency). Hence, I have moved the receiver code to `shellcmd_receive.py`. Since only the `ShellCommandSend` remains in `shellcmd.py` should I rename it to `shellcmd_send.py`?

Reply:
I think `shellcmd.py` is fine.

Reply:
see commit 0eac88a738c21c84dafecbc0d21a9511188ffaa8